jerry-core/lit/lit-strings.c

```diff
@@ -779,6 +779,57 @@ lit_code_point_to_utf8 (lit_code_point_t code_point, /**< code point */
 } /* lit_code_point_to_utf8 */
 
 /**
+ * Convert cesu-8 string to an utf-8 string and put it to the buffer.
+ * It is the caller's responsibility to make sure that the string fits in the buffer.
+ *
+ * @return number of bytes, actually copied to the utf-8 string buffer.
+ */
+lit_utf8_size_t
+lit_convert_cesu8_string_to_utf8_string (lit_utf8_byte_t *cesu8_string, /**< cesu-8 string */
+                                         lit_utf8_size_t cesu8_size, /**< size of cesu-8 string */
+                                         lit_utf8_byte_t *utf8_string, /**< destination utf-8 buffer pointer
+                                                                        * (can be NULL if buffer_size == 0) */
+                                         lit_utf8_size_t utf8_size) /**< size of utf-8 buffer */
+{
+  lit_utf8_byte_t *cesu8_pos = cesu8_string;
+  lit_utf8_byte_t *utf8_pos = utf8_string;
+
+  lit_utf8_byte_t *cesu8_end_pos = cesu8_string + cesu8_size;
+  lit_utf8_byte_t *utf8_end_pos = utf8_string + utf8_size;
+
+  lit_utf8_size_t size = 0;
+  ecma_char_t prev_ch = 0;
+
+  while (cesu8_pos < cesu8_end_pos)
+  {
+    ecma_char_t ch;
+    lit_utf8_size_t code_unit_size = lit_read_code_unit_from_utf8 (cesu8_pos, &ch);
+
+    if (lit_is_code_point_utf16_low_surrogate (ch) && lit_is_code_point_utf16_high_surrogate (prev_ch))
+    {
+      utf8_pos -= code_unit_size;
```

Comment:
This is quite clever idea.

One more thing: the cesu8 size of low and high surrogate are equals so substracting code_unit_size works, but an assert could be good here.